sway-core/src/type_system/type_binding.rs

```diff
@@ -174,18 +175,30 @@ impl TypeBinding<CallPath> {
                     errors
                 )
             }
-            TypedDeclaration::StructDeclaration(ref mut decl) => {
+            TypedDeclaration::StructDeclaration(ref original_id) => {
+                // get the copy from the declaration engine
+                let mut new_copy = check!(
+                    CompileResult::from(de_get_struct(original_id.clone(), &self.span())),
+                    return err(warnings, errors),
+                    warnings,
+                    errors
+                );
+
+                // monomorphize the copy, in place
                 check!(
                     ctx.monomorphize(
-                        decl,
+                        &mut new_copy,
                         &mut self.type_arguments,
                         EnforceTypeArguments::No,
                         &self.span
                     ),
                     return err(warnings, errors),
                     warnings,
                     errors
-                )
+                );
+
+                // insert the new copy into the declaration engine
+                de_add_monomorphized_struct_copy(original_id.clone(), new_copy);
```

Comment:
We need to return a reference to the new copy. Change this to:

```rust
                // insert the new copy into the declaration engine
                let new_id = de_insert_struct(new_copy);

                // add the new copy as a monomorphized copy of the original id
                de_add_monomorphized_copy(original_id, new_id.clone());

                TypedDeclaration::StructDeclaration(new_id)
```

See: https://github.com/FuelLabs/sway/blob/e374d6412b778df8dd70681a41bb6863265acf10/sway-core/src/type_system/type_binding.rs#L188

You may have to base this branch off of #2713, because that created the function `de_add_monomorphized_copy`.